lib/mongo/crypt/handle.rb

```diff
@@ -0,0 +1,129 @@
+
+      # Creates a new Handle object and initializes it with options
+      #
+      # @example Instantiate a Handle object with local KMS provider options
+      #   Mongo::Crypt::Handle.new({
+      #     local: { key: 'MASTER-KEY' }
+      #   })
+      #
+      # @param [ Hash ] kms_providers A hash of KMS settings. The only supported key
+      # is currently :local. Local KMS options must be passed in the format
+      # { local: { key: 'MASTER-KEY' } } where the master key is a 96-byte, base64
+      # encoded string.
+      #
+      # @since 2.12.0
+      def initialize(kms_providers)
+        @mongocrypt = Binding.mongocrypt_new
+
+        begin
+          set_kms_providers(kms_providers)
+          initialize_mongocrypt
+        rescue => e
+          self.close
+          raise e
+        end
+      end
+
+
+      # Destroy the reference to the underlying mongocrypt_t object and
+      # clean up resources
+      #
+      # @return [ true ] Always true
+      #
+      # @since 2.12.0
+      def close
+        Binding.mongocrypt_destroy(@mongocrypt) if @mongocrypt
+        @mongocrypt = nil
+
+        true
+      end
+
+      private
+
+      # Validate the kms_providers option and use it to set the KMS provider
+      # information on the underlying mongocrypt_t object
+      def set_kms_providers(kms_providers)
+        unless kms_providers
+          raise ArgumentError.new("The kms_providers option must not be blank")
+        end
+
+        unless kms_providers.key?(:local) || kms_providers.key?(:aws)
+          raise ArgumentError.new('The kms_providers option must have one of the following keys: :aws, :local')
+        end
+
+        set_kms_providers_local(kms_providers) if kms_providers.key?(:local)
+
+        if kms_providers.key?(:aws)
+          raise ArgumentError.new(':aws is not yet a supported kms_providers option. Use :local instead')
+        end
+      end
+
+      # Validate and set the local KMS provider information on the underlying
+      # mongocrypt_t object and raise an exception if the operation fails
+      def set_kms_providers_local(kms_providers)
+        unless kms_providers[:local][:key] && kms_providers[:local][:key].is_a?(String)
+          raise ArgumentError.new(
+            "The specified kms_providers option is invalid: #{kms_providers}. " +
+            "kms_providers with :local key must be in the format: { local: { key: 'MASTER-KEY' } }"
+          )
+        end
+
+        master_key = kms_providers[:local][:key]
+
+        Binary.with_binary(Base64.decode64(master_key)) do |binary|
+          success = Binding.mongocrypt_setopt_kms_provider_local(@mongocrypt, binary.ref)
+          raise_from_status unless success
+        end
+      end
+
+      # Initialize the underlying mongocrypt_t object and raise an error if the operation fails
+      def initialize_mongocrypt
+        success = Binding.mongocrypt_init(@mongocrypt)
+        # There is currently no test for this code path
```

Comment:
I think you can get this to fail based on other options you can pass in to `mongocrypt_t`, but I don't think it's possible to get to that state right now.